src/MongoDB.Driver.Core/Core/Connections/TcpStreamFactory.cs

```diff
@@ -191,12 +203,12 @@ private async Task ConnectAsync(Socket socket, EndPoint endPoint, CancellationTo
                         await Task.Factory.FromAsync(socket.BeginConnect(endPoint, null, null), socket.EndConnect).ConfigureAwait(false);
                     }
 #endif
-                    connected = true;
+                    changeState(2);
                     return;
                 }
                 catch
                 {
-                    if (!cancelled && !timedOut)
+                    if (state < 3)
```

Comment:
To be consistent, can we have the same(similar) logic as we have here: https://github.com/rstam/mongo-csharp-driver/pull/137/files#diff-ecbe46f78c404553b6544c8d7361d338R60-R66

Reply:
I don't really agree with this. It's not really clear logic from the first glance. And if it's possible to standardize it, I would suggest to do it, to save time for someone else who possible will need to understand these methods (unless it requires big changes)

Reply:
I'm more interested in fixing the race condition without doing other unnecessary refactoring that might introduce unrelated bugs or unintended behavioral changes.

Reply:
What about this case:
1. We're starting to call `Connect` which calls with delay higher than the timeout.
2. At the time when we call `Connect` and before finishing it, we trigger `socket.Dispose` based on state 3.
3. As far as I understand, It can trigger an error in executing `socket.Connect` and we will appear in the catch block which will be skipped based on line 149. 
4. As a result, we will `Dispose` socket one more time in line 157.
I don't think that it's good behavior.

But if you don't work on it in the scope of this ticket, I'm ok to create a new one.

Reply:
I'll think about it further...

But note that it's allowed to call Dispose as many times as you want.

Reply:
I've attempted a refactoring that hopefully satisfies your concerns.